### PR TITLE
Fix firewall rules on CentOS Gateway Nodes

### PR DESCRIPTION
**Pull request by sridhargaddam**
By default when iptables is installed on CentOS, it configures a REJECT rule in the FORWARDING chain.
To support inter-cluster pod-to-pod connectivity, Submariner-engine creates a new chain (i.e., SUBMARINER-FORWARD) with the necessary rules.
It then appends a rule in the FORWARD Chain to redirect the inter-cluster traffic to SUBMARINER-FORWARD chain.
Since the rule added in the FORWARD chain is appeneded at the end, REJECT rule takes priority and inter-cluster traffic does not go through.

This patch modifies the submariner code to inspect if the platform has a REJECT rule at the end of FORWARD Chain and does the following.
If there is a REJECT rule, it inserts the new rule before the REJECT rule. 
If not, it simply appends it to the end of the FORWARD Chain.

Sample output of iptables rule on CentOS7:
Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
num   pkts bytes target     prot opt in     out     source               destination         
...
X        0     0 REJECT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            reject-with icmp-host-prohibited

### CODE REVIEW

File: pkg/cableengine/ipsec/ipsec.go [hunk 1]
```diff
@@ -95,7 +95,20 @@ func (i *Engine) StartEngine(ignition bool) error {
 		forwardToSubPostroutingRuleSpec := []string{"-j", "SUBMARINER-POSTROUTING"}
 		ipt.AppendUnique("nat", "POSTROUTING", forwardToSubPostroutingRuleSpec...)
 		forwardToSubForwardRuleSpec := []string{"-j", "SUBMARINER-FORWARD"}
-		ipt.AppendUnique("filter", "FORWARD", forwardToSubForwardRuleSpec...)
+		rules, err := ipt.List("filter", "FORWARD")
+		if err != nil {
+			klog.Fatalf("error listing the rules in FORWARD chain: %v", err)
+		}
+		insertAt := 0
+		if rules[len(rules)-1] == "-A FORWARD -j REJECT --reject-with icmp-host-prohibited" {
```
[the remark this answers is not in the document]
Comment (sridhargaddam) [reply pkg/cableengine/ipsec/ipsec.go]: Thanks for reviewing @tpantelis. After some more testing, I noticed that using "Insert" can lead to duplicate entries when Submariner-Engine POD gets restarted. I'll modify the patch to avoid these duplicate entries and will consider the suggestion.

Just for notes:
If a Chain exists but has no rules, ipt.List(...) returns success and the len(rules) will be 1. The zeroth entry will be the default policy associated with the chain.

File: pkg/cableengine/ipsec/ipsec.go [hunk 2]
```diff
@@ -95,7 +95,26 @@ func (i *Engine) StartEngine(ignition bool) error {
...
+				insertAt = i
+				break
+			}
+		}
+
+		// If the FORWARD Chain does not have any REJECT rule, append the
+		// forwardToSubForwardRuleSpec rule, otherwise insert it at the appropriate location.
+		if insertAt == -1 {
+			ipt.AppendUnique("filter", "FORWARD", forwardToSubForwardRuleSpec...)
+		} else if rules[insertAt-1] != "-A FORWARD -j SUBMARINER-FORWARD" {
```
[the remark this answers is not in the document]
Comment (sridhargaddam) [reply pkg/cableengine/ipsec/ipsec.go]: I'm afraid, insertAt cannot be 0 (we would get an error when the iptables command is executed). 
Regarding the other point, I hope when you say "FORWARD", you mean "SUBMARINER-FORWARD". So, if there is already a SUBMARINER-FORWARD rule, but there is no REJECT rule, its still not a problem because we use AppendUnique. I verified this scenario.

[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/cableengine/ipsec/ipsec.go]: Re: insertAt being 0, I was referring to rules[insertAt - 1] which would blow up with an array index-out-of-bounds.  And yes by "FORWARD", I mean "-A FORWARD -j SUBMARINER-FORWARD".  Ack on the AppendUnique. Than perhaps we can just check:

    else if insertAt == 0 || rules[insertAt - 1] != "..." {
        // insert
    }

[the remark this answers is not in the document]
Comment (sridhargaddam) [reply pkg/cableengine/ipsec/ipsec.go]: As mentioned in my earlier comments, if a Chain exists (i.e., in our case, its the FORWARD chain which is one of the standard Chains which would exist) but has no rules, ipt.List(...) returns success and the len(rules) will be 1. The zeroth entry will be the default policy associated with the chain. 

So, even in such situations (i.e., empty FORWARD chain), when the following code is executed, insertAt will not be zero (since 0th entry is the default policy of the Chain and not the rules. The rules start from 1).

                for i, rule := range rules {
                        if rule == "-A FORWARD -j REJECT --reject-with icmp-host-prohibited" {
                                insertAt = i 
                                break
                        }
                }

IMHO, checking "else if insertAt == 0 || ...." is redundant. Also, if we execute ipt.Insert(...) with insertAt as 0, it will return an error.

[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/cableengine/ipsec/ipsec.go]: OK - as long as that behavior is documented and guaranteed and across all platforms etc.  I'm just playing devil's advocate here wrt strict coding, ie not assuming anything and guarding against index OOB.  It may be redundant but IMO it doesn't hurt to check (just in case...). Can you at least add a comment explaining the expected behavior of ipt and what we're assuming wrt our index logic?

[the remark this answers is not in the document]
Comment (sridhargaddam) [reply pkg/cableengine/ipsec/ipsec.go]: I see your point. Let me update the code. Thanks for reviewing.

File: pkg/cableengine/ipsec/ipsec.go [hunk 3]
```diff
@@ -95,7 +95,30 @@ func (i *Engine) StartEngine(ignition bool) error {
...
+				break
+			} else if rule == "-A FORWARD -j REJECT --reject-with icmp-host-prohibited" {
+				insertAt = i
+				break
+			}
+		}
+
+		if insertAt == appendAt {
+			// Append the rule at the end of the FORWARD Chain.
+			ipt.AppendUnique("filter", "FORWARD", forwardToSubForwardRuleSpec...)
```
Comment (tpantelis) [inline pkg/cableengine/ipsec/ipsec.go]: Since we've already checked if the entry already exists, it seems we could just use Append here to avoid the duplicate checking in AppendUnique but I assume it's relatively negligible.

File: pkg/cableengine/ipsec/ipsec.go [hunk 4]
```diff
@@ -95,7 +95,30 @@ func (i *Engine) StartEngine(ignition bool) error {
...
+		} else if insertAt > 0 {
```
Comment (tpantelis) [inline pkg/cableengine/ipsec/ipsec.go]: I don't know if it's possible for the REJECT rule to be at index 0 (I think you've stated previously that there's always or should be some default rule present) but perhaps we should at least log that case with a Warn, eg "A REJECT rule was unexpectedly found at the beginning of the FORWARD chain - the SUBMARINER-FORWARD rule was not added". Or does Insert support index 0?

Comment (sridhargaddam) [reply pkg/cableengine/ipsec/ipsec.go]: This is a sample output of FORWARD Chain.

	Chain FORWARD (policy DROP 0 packets, 0 bytes)
	num   pkts bytes target     prot opt in     out     source               destination
	1     9638   40M DOCKER-USER  all  --  *      *       0.0.0.0/0            0.0.0.0/0
	2     9638   40M DOCKER-ISOLATION-STAGE-1  all  --  *      *       0.0.0.0/0            0.0.0.0/0
	3     5238   40M ACCEPT     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0            ctstate RELATED,ESTABLISHED
	4        0     0 DOCKER     all  --  *      docker0  0.0.0.0/0            0.0.0.0/0
	5     4400  250K ACCEPT     all  --  docker0 !docker0  0.0.0.0/0            0.0.0.0/0
	6        0     0 ACCEPT     all  --  docker0 docker0  0.0.0.0/0            0.0.0.0/0
	7        0     0 REJECT     all  --  *      *       0.0.0.0/0            0.0.0.0/0            reject-with icmp-host-prohibited


ipt.List() on that host would return the following output.

    iptables_test.go:373: Rule number: 0, rule: -P FORWARD DROP 
    iptables_test.go:373: Rule number: 1, rule: -A FORWARD -j DOCKER-USER 
    iptables_test.go:373: Rule number: 2, rule: -A FORWARD -j DOCKER-ISOLATION-STAGE-1 
    iptables_test.go:373: Rule number: 3, rule: -A FORWARD -o docker0 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT 
    iptables_test.go:373: Rule number: 4, rule: -A FORWARD -o docker0 -j DOCKER 
    iptables_test.go:373: Rule number: 5, rule: -A FORWARD -i docker0 ! -o docker0 -j ACCEPT 
    iptables_test.go:373: Rule number: 6, rule: -A FORWARD -i docker0 -o docker0 -j ACCEPT 
    iptables_test.go:373: Rule number: 7, rule: -A FORWARD -j REJECT --reject-with icmp-host-prohibited 

The zero'th entry which is a default Policy (aka rule number 0) associated with the Chain, cannot have a rule like "REJECT --reject-with icmp-host-prohibited" as shown below.
So, WARN log message is not required. 

[centos@workerone iptables]$ sudo iptables -P FORWARD REJECT --reject-with icmp-host-prohibited                                                                                             
iptables v1.4.21: unknown option "--reject-with"
Try `iptables -h' or 'iptables --help' for more information.
